From 26fb556604e37aa70274dfec50022a88098a97f9 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 01 Jul 2021 16:26:33 +0300
Subject: [PATCH] [SW-47213] pci: avoid accessing hw when it's not ready

In some hosts the PRSTn signal might be asserted for a long time.
While it's asserted some hw block are held in reset and must not
be accessed.
after deassertion of the PRSTn signal the HW is performing
initialization sequence during which it's not advisable to access
it as well. (quoting Haim Revivo here)

Hence the sequense of the pci monitoring by preboot should be:
while PRSTn is low, there is no point reading any status regs. in fact
it's dangerous and might cause cpu exception.
after deassertion of the PRSTn we wait for the first link establishment,
that's our indication all the clocks are stable and HW FSMs operational.
and then we start monitoring the rest of the events (flr, hot reset,
link state, etc..)

In order to avoid duplication and race conditions, CPU1 is responsible
for the first stage of PRSTn and link monitoring.
CPU0 will monitor the running state of CPU1 and deduct the
pci state from there.
Meaning, if CPU1 has finished running (wfe state) or wasn't running
at all (reset state), then pci is initialized. otherwise pci
initialization is still in progress.

Change-Id: I4e7fe397c19c3061d1e11b875173dca10bbd49ef
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi/common.c b/arch/gaudi/common.c
index 23c0e36..d87b281 100644
--- a/arch/gaudi/common.c
+++ b/arch/gaudi/common.c
@@ -57,6 +57,8 @@
 	((BIT(1) << CPU_CA53_CFG_ARM_RST_CONTROL_NCPUPORESET_SHIFT) |	\
 	 (BIT(1) << CPU_CA53_CFG_ARM_RST_CONTROL_NCORERESET_SHIFT))
 
+#define CPU1_WFE_STAT (BIT(1) << CPU_CA53_CFG_ARM_PWR_STAT_0_STANDBYWFE_SHIFT)
+
 #define HOST_GPIO_I2C_DEBUG	24
 
 /* HOST_GPIO is set through mmPSOC_GPIO1 where each mmPSOC_GPIO serve 16 GPIOs.*/
@@ -592,3 +594,20 @@
 
 	return 0;
 }
+
+bool is_pcie_initialized(void)
+{
+	u32 rst_ctrl, wfe_stat;
+
+	rst_ctrl = regmap_readl(mmCPU_CA53_CFG_ARM_RST_CONTROL);
+	rst_ctrl &= CPU1_RST_CTRL; /* reset signals are active low */
+
+	wfe_stat = regmap_readl(mmCPU_CA53_CFG_ARM_PWR_STAT_0);
+	wfe_stat &= CPU1_WFE_STAT;
+
+	/*
+	 * if CPU1 finished running (wfe) of didn't run at all (reset state)
+	 * pcie should be in operational state.
+	 */
+	return (!rst_ctrl || wfe_stat);
+}
diff --git a/include/preboot.h b/include/preboot.h
index b5dad19..d51b49f 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -207,6 +207,7 @@
 int preboot_set_sram_sf_access(bool is_enable);
 void preboot_device_id_init(void);
 u32 preboot_device_id_get(void);
+bool is_pcie_initialized(void);
 
 #endif /* __ASSEMBLY__ */
 
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index 2987e5e..a8aebca 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -62,6 +62,7 @@
 }
 
 __weak void hl_pci_iatu_init(void) {}
+__weak bool is_pcie_initialized(void) {return true; }
 
 void module_pci_setup(void)
 {
@@ -73,6 +74,13 @@
 
 __weak enum cmd_check_res dev_specific_module_pci_check(void)
 {
+	/*
+	 * don't read pci status regs until PRSTn is deasserted and pci blocks
+	 * are functional
+	 */
+	if (!is_pcie_initialized())
+		return NOT_FOUND;
+
 	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
 	pci_sts.hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
 	pci_sts.rdlh_link_reg_val = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
